src/react/hooks/useQuery.ts

```diff
@@ -487,208 +644,246 @@ class InternalState<TData, TVariables extends OperationVariables> {
+  // we're doing server rendering and therefore override the result below.
+  if (!resultData.current) {
+    // WARNING: SIDE-EFFECTS IN THE RENDER FUNCTION
+    // this could call unsafeHandlePartialRefetch
+    setResult(
+      observable.getCurrentResult(),
+      resultData,
+      observable,
+      client,
+      partialRefetch,
+      () => {},
+      callbacks
+    );
   }
+  return resultData.current!;
+}
 
-  private toApolloError(
-    result: ApolloQueryResult<TData>
-  ): ApolloError | undefined {
-    return isNonEmptyArray(result.errors) ?
-        new ApolloError({ graphQLErrors: result.errors })
-      : result.error;
-  }
+export function getDefaultFetchPolicy<
+  TData,
+  TVariables extends OperationVariables,
+>(
+  queryHookDefaultOptions?: Partial<WatchQueryOptions<TVariables, TData>>,
+  clientDefaultOptions?: DefaultOptions
+): WatchQueryFetchPolicy {
+  return (
+    queryHookDefaultOptions?.fetchPolicy ||
+    clientDefaultOptions?.watchQuery?.fetchPolicy ||
+    "cache-first"
+  );
+}
 
-  private getCurrentResult(): ApolloQueryResult<TData> {
-    // Using this.result as a cache ensures getCurrentResult continues returning
-    // the same (===) result object, unless state.setResult has been called, or
-    // we're doing server rendering and therefore override the result below.
-    if (!this.result) {
-      this.handleErrorOrCompleted(
-        (this.result = this.observable.getCurrentResult())
-      );
-    }
-    return this.result;
-  }
+function toApolloError<TData>(
+  result: ApolloQueryResult<TData>
+): ApolloError | undefined {
+  return isNonEmptyArray(result.errors) ?
+      new ApolloError({ graphQLErrors: result.errors })
+    : result.error;
+}
 
-  // This cache allows the referential stability of this.result (as returned by
-  // getCurrentResult) to translate into referential stability of the resulting
-  // QueryResult object returned by toQueryResult.
-  private toQueryResultCache = new (canUseWeakMap ? WeakMap : Map)<
-    ApolloQueryResult<TData>,
-    QueryResult<TData, TVariables>
-  >();
-
-  toQueryResult(
-    result: ApolloQueryResult<TData>
-  ): QueryResult<TData, TVariables> {
-    let queryResult = this.toQueryResultCache.get(result);
-    if (queryResult) return queryResult;
-
-    const { data, partial, ...resultWithoutPartial } = result;
-    this.toQueryResultCache.set(
-      result,
-      (queryResult = {
-        data, // Ensure always defined, even if result.data is missing.
-        ...resultWithoutPartial,
-        ...this.obsQueryFields,
-        client: this.client,
-        observable: this.observable,
-        variables: this.observable.variables,
-        called: !this.queryHookOptions.skip,
-        previousData: this.previousData,
-      })
-    );
+export function toQueryResult<TData, TVariables extends OperationVariables>(
+  result: ApolloQueryResult<TData>,
+  resultData: InternalResult<TData, TVariables>,
```

Comment:
```suggestion
  previousData: TData | undefined,
```

Could we pass in `previousData` here instead? The `resultData.current = toQueryResult(resultData)` feels a tad odd because I'd expect to mutate `resultData` or something else. Changing this to `previousData` makes it more obvious what the dependency is here.